src/ray/common/asio/io_context.cc

```diff
@@ -0,0 +1,257 @@
+    // this allows the common path, in which the handler already exists in the hash table,
+    // to only require the readers lock.
+    absl::WriterMutexLock lock(&mutex_);
+    const auto pair =
+        post_handler_stats_.try_emplace(name, std::make_shared<GuardedHandlerStats>());
+    if (pair.second) {
+      it = pair.first;
+    } else {
+      it = post_handler_stats_.find(name);
+      // If try_emplace failed to insert the item, the item is guaranteed to exist in
+      // the table.
+      RAY_CHECK(it != post_handler_stats_.end());
+    }
+  }
+  {
+    absl::WriterMutexLock lock(&(it->second->mutex));
+    it->second->stats.cum_count++;
+    it->second->stats.curr_count++;
+  }
+  const auto start = std::chrono::steady_clock::now();
+  // References are only invalidated upon deletion of the corresponding item from the
+  // table, which we won't do until this io_context is deleted. Provided that
+  // GuardedHandlerStats synchronizes internal access, we can concurrently write to these
+  // stats from multiple threads without acquiring a table-level readers lock in the
+  // callback.
+  auto &stats = it->second;
+  // TODO(Clark): Use generalized lambda capture to move `handler` into the lambda once
+  // we move to C++14.
+  auto wrapped_handler = capture(
+      std::move(handler), [this, name, start, stats](std::function<void()> handler) {
+        const auto start_execution = std::chrono::steady_clock::now();
+        // Execute actual handler.
+        handler();
+        const auto end_execution = std::chrono::steady_clock::now();
+        // Update queue and execution time stats.
+        const auto queue_time_ns =
+            std::chrono::duration_cast<std::chrono::nanoseconds>(start_execution - start)
+                .count();
+        const auto execution_time_ns =
+            std::chrono::duration_cast<std::chrono::nanoseconds>(end_execution -
+                                                                 start_execution)
+                .count();
+        // Update handler-specific stats.
+        {
+          absl::WriterMutexLock lock(&(stats->mutex));
+          // Handler-specific queue stats.
+          stats->stats.cum_queue_time += queue_time_ns;
+          if (stats->stats.min_queue_time > queue_time_ns) {
+            stats->stats.min_queue_time = queue_time_ns;
+          }
+          if (stats->stats.max_queue_time < queue_time_ns) {
+            stats->stats.max_queue_time = queue_time_ns;
+          }
+          // Handler-specific execution stats.
+          stats->stats.cum_execution_time += execution_time_ns;
+          if (stats->stats.min_execution_time > execution_time_ns) {
+            stats->stats.min_execution_time = execution_time_ns;
+          }
+          if (stats->stats.max_execution_time < execution_time_ns) {
+            stats->stats.max_execution_time = execution_time_ns;
+          }
+          // Handler-specific current count.
+          stats->stats.curr_count--;
```

Comment:
How about we only store handler-specific stats for total execution time (drop queueing, and exec min/max)? I don't think queueing is meaningful on a per-handler basis. Min/max is probably overkill for our purposes.

Reply:
Right now, execution time doesn't include queueing time. Should execution time instead be measured from the time at which the handler is enqueued?

Reply:
The existing definition seems fine. Queueing is just not useful to measure at all at the handler level since it's an aggregate effect. You should get similar info from just global q stats.

Reply:
I agree, although I did find it useful as an indicator of "which handler is getting stuck in the queue for longer than it should".